xline/src/storage/index.rs

```diff
@@ -134,47 +112,57 @@ pub(super) trait IndexOperate {
 
 impl IndexOperate for Index {
     fn get(&self, key: &[u8], range_end: &[u8], revision: i64) -> Vec<Revision> {
-        let inner = self.inner.lock();
         match RangeType::get_range_type(key, range_end) {
-            RangeType::OneKey => inner
-                .index
-                .get(key)
-                .and_then(|revs| Self::get_revision(revs, revision))
-                .map(|rev| vec![rev])
-                .unwrap_or_default(),
-            RangeType::AllKeys => inner
-                .index
-                .values()
-                .filter_map(|revs| Self::get_revision(revs, revision))
-                .collect(),
-            RangeType::Range => inner
-                .index
-                .range(KeyRange::new(key, range_end))
-                .filter_map(|(_k, revs)| Self::get_revision(revs, revision))
-                .collect(),
+            RangeType::OneKey => self.inner.map_read(|inner| {
+                inner
+                    .index
+                    .get(key)
+                    .and_then(|revs| Self::get_revision(revs, revision))
+                    .map(|rev| vec![rev])
+                    .unwrap_or_default()
+            }),
+            RangeType::AllKeys => self.inner.map_read(|inner| {
+                inner
+                    .index
+                    .values()
+                    .filter_map(|revs| Self::get_revision(revs, revision))
+                    .collect()
+            }),
+            RangeType::Range => self.inner.map_read(|inner| {
+                inner
+                    .index
+                    .range(KeyRange::new(key, range_end))
+                    .filter_map(|(_k, revs)| Self::get_revision(revs, revision))
+                    .collect()
+            }),
         }
```

Comment:
Please move `map_read` to the outside of the `match` block to reduce the duplication

Reply:
The `get` method is a hotspot. We need to keep the critical section as small as possible to minimize lock contention as possible.  The `match` block and `get_range_type` methods do not require holding the lock, which is why I removed them from the `map_read` block. As for reducing duplication, I tried making the match block return a closure - `FnOnce(RwLockReadGuard<'_, IndexInner>) -> Vec<Revision>`. This would reduce some duplication but would require using `Box<dyn FnOnce(RwLockReadGuard<'_, IndexInner>) -> Vec<Revision>>` to wrap the closure, which would result in a performance loss. After weighing the pros and cons, I still believe that using this approach is better. Although there is some seemingly duplicated code, the readability is not that bad. Additionally, the type of Key appears to be relatively stable and unlikely to change, so it is less likely that a modification in one place would be forgotten to be modified in another copy. Any thoughts of this? @rogercloud